deltacat/compute/stats/utils/intervals.py

```diff
@@ -1,62 +1,88 @@
-    a minimal number of closed bounded output intervals in O(N log N) time.
+    Each input interval is an open unbounded interval with infinite ranges.
+    Intervals will be merged into a minimal number of output intervals in O(N log N) time.
 
     Example:
         >>> merge_intervals((3, 9), (8, 12), (15, 19))
         ((3, 12), (15, 19))
 
+    Example:
+        >>> merge_intervals((3, 9), (None, 15), (13, 30))
+        ((None, 30))
+
     Args:
-        intervals: A list of closed bounded intervals
+        intervals: A list of open bounded intervals with an int type representing finite, closed bounded values and
+            a None type representing infinity.
 
     Returns:
-        A minimal number of closed bounded output intervals
+        A minimal number of output intervals
     """
-    merged = set()
-    intervals_list: List[Tuple[int, int]] = list(intervals)
+    merged: Set[Tuple[UnboundedRange, UnboundedRange]] = set()
+    intervals_list: List[Tuple[UnboundedRange, UnboundedRange]] = list(intervals)
 
     if len(intervals_list) == 0:
         return merged
 
+    _to_numeric_values(intervals_list)
     intervals_list.sort()  # sort by starting range numbers
 
     merge_start, merge_end = None, None
     for interval in intervals_list:
-        start, end = _get_validated_interval(interval)
+        start, end = interval
+        if start > end:
+            raise ValueError(f"Invalid stream position range interval: ({start}, {end})")
 
         if merge_start is None and merge_end is None:
             merge_start, merge_end = start, end
             continue
 
         if merge_end < start:
-            merged.add((merge_start, merge_end))  # add current merge interval if no overlap, begin new interval
+            # add current merge interval if no overlap, begin new interval
+            _add_merged_interval(merged, merge_start, merge_end)
             merge_start, merge_end = start, end
         elif merge_end < end:
             merge_end = end  # expand current merge interval if there is an overlap
 
-    merged.add((merge_start, merge_end))  # add final merge interval
+    # add final merge interval
+    _add_merged_interval(merged, merge_start, merge_end)
 
     return merged
 
 
-def _get_validated_interval(interval: Tuple[int, int]) -> Tuple[int, int]:
+def _add_merged_interval(result_set: set, start: NumericalRange, end: NumericalRange):
+    start_pos: UnboundedRange = start if isinstance(start, int) else None
+    end_pos: UnboundedRange = end if isinstance(end, int) else None
+    result_set.add((start_pos, end_pos))
+
+
+def _to_numeric_values(intervals_list: List[Tuple[UnboundedRange, UnboundedRange]]):
+    for i, interval in enumerate(intervals_list):
+        start, end = _get_validated_interval(interval)
+        if start is None:
+            start = float('-inf')
+        if end is None:
+            end = float('inf')
+
+        intervals_list[i] = (start, end)
```

Comment:
Let's drop this together with our support for `math.inf`.

Reply:
I think we'll still need this conversion to sort and merge the intervals for a given None type value

Reply:
The `merge_intervals` function internally converts `None` values to numerical values, to make use of the sort. After the merging process, the float (inf) types are converted back to `None`. This is done to be compliant with the `list_deltas` API, which accepts `Optional[int]` for starting and ending stream positions. 

So I think we could keep the `NumericalRange` type for internal conversions (None/infinity) but keep the function input/output types to be a simple `Optional[int]`.